### PR TITLE
Bug 1802534: gcp-routes: move to MCO, implement downfile, tweak timing

### PR DESCRIPTION
**Pull request by squeed**
Move gcp-routes to the MCO, call it openshift-gcp-routes. Also, improve the behavior significantly:

1. always accept connections from external sources
2. use downfiles instead of stopping the service
3. implement the equivalent in gcp-routes-controller.

### CODE REVIEW

**Comment by squeed** [comment]
cc @sttts

File: cmd/gcp-routes-controller/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# gcp-routes-controller
+
+## Background
+
+Google cloud load balancer is a L3LB that is Special. It doesn't do DNAT; instead, it
```
Comment (sttts) [inline cmd/gcp-routes-controller/README.md:5]: nit: special

File: cmd/gcp-routes-controller/README.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+just redirects traffic to backends and preserves the VIP as the destination IP.
+
+So, an agent exists on the node. It programs the node (either via iptables or routing tables) to
+accept traffic destined for the VIP. However, this has a problem: all hairpin traffic
+to the balanced servce is *always* handled by that backend.
```
Comment (sttts) [inline cmd/gcp-routes-controller/README.md:10]: ... even if it is down or unhealthy.

**Comment by squeed** [comment]
PR adding downfile support to gcp-routes.sh: https://gitlab.cee.redhat.com/coreos/redhat-coreos/merge_requests/899

File: cmd/gcp-routes-controller/run.go [hunk 3]
```diff
@@ -82,15 +97,14 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
...
-
 	errCh := make(chan error)
 	tracker := &healthTracker{
 		state:            unknownTrackerState,
 		ErrCh:            errCh,
 		SuccessThreshold: 2,
-		FailureThreshold: 10,
-		OnFailure:        func() error { return exec.Command("systemctl", "stop", runOpts.gcpRoutesService).Run() },
-		OnSuccess:        func() error { return exec.Command("systemctl", "start", runOpts.gcpRoutesService).Run() },
+		FailureThreshold: 5, // it takes 20 seconds for vip changes to propagate, so give us 25 seconds before we react
```
Comment (sttts) [inline cmd/gcp-routes-controller/run.go:105]: would even go down to 3

File: cmd/gcp-routes-controller/run.go [hunk 4]
```diff
@@ -129,6 +142,90 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
...
+	} else {
+		glog.Infof("%s not accessible, will stop gcp-routes.service on health failure", realPath)
+		h.mode = modeStopService
+	}
+
+	// if StopService mode and rootfs specified, chroot
+	if h.mode == modeStopService && runOpts.rootMount != "" {
+		glog.Infof(`Calling chroot("%s")`, runOpts.rootMount)
+		if err := syscall.Chroot(runOpts.rootMount); err != nil {
+			return nil, fmt.Errorf("Unable to chroot to %s: %s", runOpts.rootMount, err)
```
Comment (sttts) [inline cmd/gcp-routes-controller/run.go:163]: nit: lower-case go errors

File: templates/master/00-master/gcp/files/etc-kubernetes-manifests-gcp-routes-controller.yaml [hunk 5]
```diff
@@ -15,7 +15,7 @@ contents:
         command: ["gcp-routes-controller"]
         args:
         - "run"
-        - "--health-check-url=https://127.0.0.1:6443/readyz"
+        - "--health-check-url={{.Infra.Status.APIServerInternalURL}}/readyz"
```
Comment (sttts) [inline templates/master/00-master/gcp/files/etc-kubernetes-manifests-gcp-routes-controller.yaml:18]: just to be sure, is this written once or dynamically updated? Afaik in theory APIServerInternalURL could change?

Comment (squeed) [reply templates/master/00-master/gcp/files/etc-kubernetes-manifests-gcp-routes-controller.yaml:18]: I don't think anything supports this changing. AFAICT the "status" is generated in the installer.

**Comment by squeed** [comment]
/hold
until the gcp PR is merged and "deployed" in to CI.

File: templates/master/00-master/gcp/files/etc-kubernetes-manifests-gcp-routes-controller.yaml (hunk 5 above)
Comment (sttts) [reply templates/master/00-master/gcp/files/etc-kubernetes-manifests-gcp-routes-controller.yaml:18]: ok. then it's fine.

File: cmd/gcp-routes-controller/README.md (hunk 1 above)
Comment (squeed) [reply cmd/gcp-routes-controller/README.md:5]: hah, needless capitalization is classic Nerd English but I'll change it :p

File: cmd/gcp-routes-controller/run.go (hunk 3 above)
Comment (squeed) [reply cmd/gcp-routes-controller/run.go:105]: yeah, fair enough; it's just kubelet, sdn, and multus that use this.

**Comment by squeed** [comment]
Seems the unit test scaffolding is out-of-date, so this is not going to pass unit tests. Investigating.

**Comment by squeed** [comment]
Seems the unit test scaffolding is out-of-date, so this is not going to pass unit tests. Investigating.

File: cmd/gcp-routes-controller/run.go [hunk 6]
```diff
@@ -111,11 +125,10 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
...
 		for sig := range c {
 			glog.Infof("Signal %s received: shutting down gcp routes service", sig)
-			if err := exec.Command("systemctl", "stop", runOpts.gcpRoutesService).Run(); err != nil {
-				glog.Infof("Failed to terminate gcp routes service on signal: %s", err)
-			} else {
-				break
+			if err := handler.onFailure(); err != nil {
+				glog.Infof("Failed to mark service down on signal: %s", err)
 			}
+			os.Exit(0)
```
Comment (tedyu) [inline cmd/gcp-routes-controller/run.go:131]: Should this be in the else block (err == nil) ?

File: cmd/gcp-routes-controller/run.go [hunk 7]
```diff
@@ -129,6 +142,90 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
...
+	}
+
+	// otherwise, resolve vip
+	if h.mode == modeDownFile {
+		if runOpts.vip != "" {
+			h.vip = runOpts.vip
+		} else {
+			addrs, err := net.LookupHost(uri.Hostname())
+			if err != nil {
+				return nil, fmt.Errorf("failed to lookup host: %v", err)
```
Comment (tedyu) [inline cmd/gcp-routes-controller/run.go:179]: uri.Hostname() should be included in the message

File: cmd/gcp-routes-controller/run.go [hunk 8]
```diff
@@ -129,6 +142,90 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
...
+			}
+			if len(addrs) != 1 {
+				return nil, fmt.Errorf("hostname %s has %d addresses, expected 1 - aborting", uri.Hostname(), len(addrs))
```
Comment (tedyu) [inline cmd/gcp-routes-controller/run.go:182]: Looking at line 185 below, we are not aborting

File: cmd/gcp-routes-controller/run.go [hunk 9]
```diff
@@ -129,6 +142,90 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
...
+// onFailure: either stop the routes service, or write downfile
+func (h *handler) onFailure() error {
+	if h.mode == modeDownFile {
+		downFile := path.Join(runOpts.rootMount, downFileDir, fmt.Sprintf("%s.down", h.vip))
+		fp, err := os.OpenFile(downFile, os.O_CREATE, 0644)
+		if err != nil {
+			return fmt.Errorf("failed to create downfile (%s): %v", downFile, err)
+		}
+		_ = fp.Close()
+		glog.Infof("healthcheck failed, created %s", downFile)
```
Comment (tedyu) [inline cmd/gcp-routes-controller/run.go:201]: It would make log searching easier if 'downFile' is included in the message (as is done on line 198 above)

**Comment by kikisdeliveryservice** [comment]
seeing:
`E0421 19:17:08.844158       1 leaderelection.go:331] error retrieving resource lock openshift-sdn/openshift-network-controller: Get https://api-int.ci-op-jd83nzz1-1354f.origin-ci-int-gce.dev.openshift.com:6443/api/v1/namespaces/openshift-sdn/configmaps/openshift-network-controller: dial tcp 10.0.0.2:6443: i/o timeout`

https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/1670/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1940/artifacts/e2e-gcp-op/pods/openshift-sdn_sdn-controller-2t46q_sdn-controller.log

**Comment by squeed** [comment]
> seeing:
> `E0421 19:17:08.844158 1 leaderelection.go:331] error retrieving resource lock openshift-sdn/openshift-network-controller: Get https://api-int.ci-op-jd83nzz1-1354f.origin-ci-int-gce.dev.openshift.com:6443/api/v1/namespaces/openshift-sdn/configmaps/openshift-network-controller: dial tcp 10.0.0.2:6443: i/o timeout`
> 
> https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/1670/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1940/artifacts/e2e-gcp-op/pods/openshift-sdn_sdn-controller-2t46q_sdn-controller.log

Nuts. Investigating.

**Comment by squeed** [comment]
Those error messages are red herrings; the node maintained access to the internal api server at that time. I suspect they're just from masters rebooting. Continuing the investigation.

**Comment by squeed** [comment]
Okay, this is weird: tests are failing because one of the worker nodes failed to start the MCD after a reboot - because CRIO won't create the sandbox.

`Apr 21 21:21:32.525864 ci-op-d4rtc-w-b-xrmnd.c.openshift-gce-devel-ci.internal hyperkube[1630]: I0421 21:21:32.525753    1630 event.go:278] Event(v1.ObjectReference{Kind:"Pod", Namespace:"openshift-machine-config-operator", Name:"machine-config-daemon-4kr98", UID:"fb999c1d-c1ac-4e8f-bf2e-98e2b785acf4", APIVersion:"v1", ResourceVersion:"27971", FieldPath:""}): type: 'Warning' reason: 'FailedCreatePodSandBox' Failed to create pod sandbox: rpc error: code = Unknown desc = reserving pod sandbox name: error reserving pod name k8s_machine-config-daemon-4kr98_openshift-machine-config-operator_fb999c1d-c1ac-4e8f-bf2e-98e2b785acf4_1 for id 6bfb87bd099b9954b3ee5654c35e9b6fc8403df3f06f6f1a57331195f94dc28e: name is reserved`

This definitely isn't caused by my change - mine only touches the master nodes. But it's not good. I wonder if the problem lies with kubelet, crio, or libpod?

**Comment by squeed** [comment]
Known (and very scary) bug: 1785399
/retest

File: cmd/gcp-routes-controller/run.go (hunk 6 above)
Comment (squeed) [reply cmd/gcp-routes-controller/run.go:131]: I don't think so - this is in the SIGINT handler. Kubelet is asking us to shut down and go away.

File: cmd/gcp-routes-controller/run.go (hunk 8 above)
Comment (squeed) [reply cmd/gcp-routes-controller/run.go:182]: Not sure what you mean - we return err and handle that on line 81?

File: cmd/gcp-routes-controller/run.go (hunk 9 above)
Comment (squeed) [reply cmd/gcp-routes-controller/run.go:201]: good point, updated.

**Comment by squeed** [comment]
Updated, fixed unit test scaffolding. This job is pretty prone to triggering the crio leak, so it might be a while before this goes green. AFAICT it's not a bug in the PR.

File: cmd/gcp-routes-controller/run.go (hunk 8 above)
Comment (tedyu) [reply cmd/gcp-routes-controller/run.go:182]: If we abort, is the "Using VIP %s" log necessary ?

**Comment by mrunalp** [comment]
@squeed opened https://bugzilla.redhat.com/show_bug.cgi?id=1826896 to fix the issue you ran into.

**Comment by kikisdeliveryservice** [comment]
> /hold
> until the gcp PR is merged and "deployed" in to CI.

@squeed has this other PR merged yet?

**Comment by sttts** [comment]
> @squeed has this other PR merged yet?

It has: https://gitlab.cee.redhat.com/coreos/redhat-coreos/merge_requests/899

/lgtm
/approve

**Comment by sttts** [comment]
/hold

checking whether https://gitlab.cee.redhat.com/coreos/redhat-coreos/merge_requests/899 is in promoted image

**Comment by sttts** [comment]
For the record, don't merge this. With the current gcp-routes implementation it makes the situation worse.

**Comment by squeed** [comment]
@sttts fixed the issues with the iptables scaffolding. When https://gitlab.cee.redhat.com/coreos/redhat-coreos/merge_requests/905 merges (and builds in to CI), then this should be good to go.

File: cmd/gcp-routes-controller/run.go [hunk 10]
```diff
@@ -82,15 +97,14 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
...
-
 	errCh := make(chan error)
 	tracker := &healthTracker{
 		state:            unknownTrackerState,
 		ErrCh:            errCh,
 		SuccessThreshold: 2,
-		FailureThreshold: 10,
-		OnFailure:        func() error { return exec.Command("systemctl", "stop", runOpts.gcpRoutesService).Run() },
-		OnSuccess:        func() error { return exec.Command("systemctl", "start", runOpts.gcpRoutesService).Run() },
+		FailureThreshold: 3, // it takes at most 20 seconds for vip changes to propagate
```
Comment (sttts) [inline cmd/gcp-routes-controller/run.go:105]: please add more comments here why these values are crucial.

Comment (squeed) [reply cmd/gcp-routes-controller/run.go]: done, and poked timings.

File: cmd/gcp-routes-controller/run.go [hunk 11]
```diff
@@ -82,22 +97,27 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
...
+
+	// careful: the timing here needs to correspond to the load balancer's
+	// parameters. We need to remove routes just after we've been removed
+	// as a backend in the load-balancer, and add routes before we've been
+	// re-added.
+	// see openshift/installer/data/data/gcp/network/lb-private.tf
 	tracker := &healthTracker{
 		state:            unknownTrackerState,
 		ErrCh:            errCh,
 		SuccessThreshold: 2,
```
[the remark this answers is not in the document]
Comment (squeed) [reply cmd/gcp-routes-controller/run.go:110]: Is there any chance the apiserver can flap on startup? If not, then 1 is good.

**Comment by squeed** [comment]
Okay, I've updated this to create an openshift-specific service and move it in to the MCO.

**Comment by LorbusChris** [comment]
If we want to pull this out of the base OS and into MCO, I still wonder how are we going to run this on the bootstrap node?

**Comment by squeed** [comment]
> If we want to pull this out of the base OS and into MCO, I still wonder how are we going to run this on the bootstrap node?

We'll have to keep a generic version in the base OS, without all the openshift customizations. Eventually that can move to nm-cloud-setup.

**Comment by vrutkovs** [comment]
>I still wonder how are we going to run this on the bootstrap node?

bootstrap node could extract this file from MCO image? 

Perhaps its worth investing into `machine-config-daemon prepare-boostrap-node` subcommand which would set GCP routes file, update CRI-O config etc.

**Comment by squeed** [comment]
Whelp, I can't get the changes to actually create a bootable master. And without the serial console logs, I'm not sure what the issue is.

**Comment by LorbusChris** [comment]
There are no logs from the control-planes machines at all. This looks weird:
from `log-bundle-20200513194307/unit-status/afterburn-sshkeys@core.service.txt`:
```
● afterburn-sshkeys@core.service - Afterburn (SSH Keys)
   Loaded: loaded (/usr/lib/systemd/system/afterburn-sshkeys@.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Wed 2020-05-13 18:57:56 UTC; 45min ago
  Process: 1658 ExecStart=/usr/bin/afterburn ${AFTERBURN_OPT_PROVIDER} --ssh-keys=core (code=exited, status=1/FAILURE)
 Main PID: 1658 (code=exited, status=1/FAILURE)

May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal afterburn[1658]: May 13 18:57:56.513 INFO Fetching http://metadata.google.internal/computeMetadata/v1/project/attributes/sshKeys: Attempt #1
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal afterburn[1658]: May 13 18:57:56.514 INFO Fetch successful
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal afterburn[1658]: May 13 18:57:56.515 INFO Fetching http://metadata.google.internal/computeMetadata/v1/project/attributes/ssh-keys: Attempt #1
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal afterburn[1658]: May 13 18:57:56.518 INFO Fetch successful
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal afterburn[1658]: Error: writing ssh keys
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal afterburn[1658]: Caused by: invalid key format
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal afterburn[1658]: Caused by: Invalid last symbol 86, offset 366.
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal systemd[1]: afterburn-sshkeys@core.service: Main process exited, code=exited, status=1/FAILURE
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal systemd[1]: afterburn-sshkeys@core.service: Failed with result 'exit-code'.
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal systemd[1]: Failed to start Afterburn (SSH Keys).
```
and in `.openshift_install.log`:
```
time="2020-05-13T19:43:20Z" level=debug msg="Gather remote logs"
time="2020-05-13T19:43:20Z" level=debug msg="Collecting info from 10.0.0.5"
time="2020-05-13T19:43:20Z" level=debug msg="lost connection"
time="2020-05-13T19:43:20Z" level=debug msg="ssh: connect to host 10.0.0.5 port 22: Connection refused\r"
time="2020-05-13T19:43:20Z" level=debug msg="Collecting info from 10.0.0.3"
time="2020-05-13T19:43:20Z" level=debug msg="lost connection"
time="2020-05-13T19:43:20Z" level=debug msg="ssh: connect to host 10.0.0.3 port 22: Connection refused\r"
time="2020-05-13T19:43:20Z" level=debug msg="Collecting info from 10.0.0.4"
time="2020-05-13T19:43:20Z" level=debug msg="lost connection"
time="2020-05-13T19:43:20Z" level=debug msg="ssh: connect to host 10.0.0.4 port 22: Connection refused\r"
```

**Comment by LorbusChris** [comment]
maybe this was a flake?^ Other e2e tests run fine.
/test e2e-gcp-op

**Comment by squeed** [comment]
/test e2e-gcp-op
so I can grab some proposed ignition config.

**Comment by squeed** [comment]
Aha!

`CRITICAL : Ignition failed: failed to create files: failed to create files: failed to create file "/usr/libexec/openshift-gcp-routes.sh": unlinkat /sysroot/usr/libexec/openshift-gcp-routes.sh: read-only file system`

**Comment by squeed** [comment]
Jumped on to a node, saw that everything is working as planned - rhcos-provided service is disabled, new service up and running, iptables looks correct.

File: cmd/gcp-routes-controller/run.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (sttts) [reply cmd/gcp-routes-controller/run.go:110]: I think when it is ready it is ready. No reason to expect flapping.

**Comment by squeed** [comment]
Tweaked timing based on sttts' suggestion; updated.

**Comment by squeed** [comment]
those aren't real failures, they're cancelled jobs because of the force-push.

**Comment by squeed** [comment]
e2e-gcp-op passed. I'm satisfied with this now. This is ready for final review.

**Comment by ashcrow** [comment]
cc @runcom

**Comment by kikisdeliveryservice** [comment]
@runcom has already approved this PR

deferring to @sttts for final LGTM

**Comment by ashcrow** [comment]
> @runcom has already approved this PR
> 
> deferring to @sttts for final LGTM

Ahh it was hidden in the 25 rolled up messages.

**Comment by squeed** [comment]
This seems to be a horrid test issue.
/hold
to stop the retests until it clears up

**Comment by squeed** [comment]
/hold cancel
now that quay is back
/retest

**Comment by squeed** [comment]
It looks like whatever the problem in https://bugzilla.redhat.com/show_bug.cgi?id=1828606 is biting us.

Given that this doesn't touch aws at all, would this be a candidate for an /override ?

**Pull request merged**